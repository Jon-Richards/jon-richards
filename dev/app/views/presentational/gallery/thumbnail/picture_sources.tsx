import * as React from 'react';

/**
 * Renders an array of <source> tags.  There are ways to produce this behavior
 * with a single srcset attribute, but this method offers better readability at
 * runtime.
 */
export function PictureSources(props: PictureSourcesProps): JSX.Element {
  const mapped = props.sources.map(source => (
    <source
      key={source.source}
      media={source.mediaQuery}
      srcSet={source.source}
      data-testid="gallery__thumb__picture-source"
    />
  ));

  const reversed = reverseForMobileFirst(mapped);

  return <React.Fragment>{reversed}</React.Fragment>;
}

export type PictureSourcesProps = {
  sources: PictureSource[];
}

type PictureSource = {
  mediaQuery: string;
  source: string;
}

/**
 * The browser will use whichever <source> tag matches the environment first,
 * so we can't assume that "min-width" will cascade the way it does with CSS.
 * (If the first source represents the smallest width, it will always return
 * true.) This function simply reverses the array so that larger resolutions
 * will be checked first.
 */
function reverseForMobileFirst(sources: JSX.Element[]): JSX.Element[] {
  return sources.reverse();
}
