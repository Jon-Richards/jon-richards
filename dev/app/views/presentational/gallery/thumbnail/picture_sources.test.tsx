import * as React from 'react';
import { PictureSources, PictureSourcesProps } from './picture_sources';
import { render } from 'enzyme';

describe('The PictureSources component.', () => {
  it('Should match the snapshot.', () => {
    const sources: PictureSourcesProps['sources'] = [
      { mediaQuery: 'min-width(320px)', source: 'test_small.jpg' },
      { mediaQuery: 'min-width(480px)', source: 'test_medium.jpg' },
      { mediaQuery: 'min-width(720px)', source: 'test_large.jpg' },
    ];
    const test_1 = render(<PictureSources sources={sources} />);

    expect(test_1).toMatchSnapshot();
  });

  it('Renders a source tag for each provided source.', () => {
    const sources: PictureSourcesProps['sources'] = [
      { mediaQuery: 'min-width(320px)', source: 'test_small.jpg' },
      { mediaQuery: 'min-width(480px)', source: 'test_medium.jpg' },
      { mediaQuery: 'min-width(720px)', source: 'test_large.jpg' },
    ];
    const test_1 = render(<PictureSources sources={sources} />);

    expect(test_1.length).toBe(3);
  });
});
