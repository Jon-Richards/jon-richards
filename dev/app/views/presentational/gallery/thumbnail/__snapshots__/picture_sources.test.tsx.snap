// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The PictureSources component. Should match the snapshot. 1`] = `
Array [
  <source
    data-testid="gallery__thumb__picture-source"
    media="min-width(720px)"
    srcset="test_large.jpg"
  />,
  <source
    data-testid="gallery__thumb__picture-source"
    media="min-width(480px)"
    srcset="test_medium.jpg"
  />,
  <source
    data-testid="gallery__thumb__picture-source"
    media="min-width(320px)"
    srcset="test_small.jpg"
  />,
]
`;
